crates/alloy-dyn-contract/src/call.rs

```diff
@@ -0,0 +1,187 @@
+use crate::error::{Error, Result};
+use alloy_dyn_abi::{DynSolValue, FunctionExt};
+use alloy_json_abi::Function;
+use alloy_primitives::{Address, Bytes, U256, U64};
+use alloy_providers::provider::TempProvider;
+use alloy_rpc_types::{state::StateOverride, BlockId, CallInput, CallRequest};
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+};
+
+/// A builder for sending a transaction via. `eth_sendTransaction`, or calling a function via
+/// `eth_call`.
+///
+/// The builder can be `.await`ed directly which is equivalent to invoking [`CallBuilder::call`].
+///
+/// # Note
+///
+/// Sets the [state overrides](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set) for `eth_call`, but this is not supported by all clients.
+pub struct CallBuilder<P> {
+    // todo: this will not work with `send_transaction` and does not differentiate between EIP-1559
+    // and legacy tx
+    request: CallRequest,
```

Comment:
I think having access to the CallRequest would be ideal. This way we can get around not having `send_transaction` on the provider by transferring the fields manually to a signable/rlp-able tx type from alloy consensus e.g `TxLegacy`